src/code/audio/audio_heap.c

```diff
@@ -1,111 +1,1576 @@
+
+    AudioHeap_ResetLoadStatus();
+    gAudioContext.notes =
+        AudioHeap_AllocZeroed(&gAudioContext.notesAndBuffersPool, gAudioContext.numNotes * sizeof(Note));
+    AudioPlayback_NoteInitAll();
+    AudioPlayback_InitNoteFreeList();
+    gAudioContext.noteSubsEu =
+        AudioHeap_AllocZeroed(&gAudioContext.notesAndBuffersPool, gAudioContext.audioBufferParameters.updatesPerFrame *
+                                                                      gAudioContext.numNotes * sizeof(NoteSubEu));
+
+    for (j = 0; j != 2; j++) {
+        gAudioContext.abiCmdBufs[j] = AudioHeap_AllocDmaMemoryZeroed(&gAudioContext.notesAndBuffersPool,
+                                                                     gAudioContext.maxAudioCmds * sizeof(u64));
+    }
+
+    gAudioContext.unk_3520 = AudioHeap_Alloc(&gAudioContext.notesAndBuffersPool, 0x100 * sizeof(f32));
+    func_8018B10C();
+    for (i = 0; i < ARRAY_COUNT(gAudioContext.synthesisReverbs); i++) {
+        gAudioContext.synthesisReverbs[i].useReverb = 0;
+    }
+
+    gAudioContext.numSynthesisReverbs = spec->numReverbs;
+    for (i = 0; i < gAudioContext.numSynthesisReverbs; i++) {
+        AudioHeap_InitReverb(i, &spec->reverbSettings[i], 1);
+    }
+
+    AudioSeq_InitSequencePlayers();
+    for (j = 0; j < gAudioContext.audioBufferParameters.numSequencePlayers; j++) {
+        AudioSeq_InitSequencePlayerChannels(j);
+        AudioSeq_ResetSequencePlayer(&gAudioContext.seqPlayers[j]);
+    }
+
+    AudioHeap_InitSampleCaches(spec->persistentSampleCacheMem, spec->temporarySampleCacheMem);
+    AudioLoad_InitSampleDmaBuffers(gAudioContext.numNotes);
+    gAudioContext.preloadSampleStackTop = 0;
+    AudioLoad_InitSlowLoads();
+    AudioLoad_InitScriptLoads();
+    AudioLoad_InitAsyncLoads();
+    gAudioContext.unk_4 = 0x1000;
+    AudioLoad_LoadPermanentSamples();
+    intMask = osSetIntMask(1);
+    osWritebackDCacheAll();
+    osSetIntMask(intMask);
+}
+
+void* AudioHeap_SearchPermanentCache(s32 tableType, s32 id) {
+    s32 i;
+
+    for (i = 0; i < gAudioContext.permanentPool.count; i++) {
+        if (gAudioContext.permanentCache[i].tableType == tableType && gAudioContext.permanentCache[i].id == id) {
+            return gAudioContext.permanentCache[i].ptr;
+        }
+    }
+    return NULL;
+}
+
+void* AudioHeap_AllocPermanent(s32 tableType, s32 id, u32 size) {
+    void* ret;
+    s32 index;
+
+    index = gAudioContext.permanentPool.count;
+
+    ret = AudioHeap_Alloc(&gAudioContext.permanentPool, size);
+    gAudioContext.permanentCache[index].ptr = ret;
+    if (ret == NULL) {
+        return NULL;
+    }
+    gAudioContext.permanentCache[index].tableType = tableType;
+    gAudioContext.permanentCache[index].id = id;
+    gAudioContext.permanentCache[index].size = size;
+    //! @bug UB: missing return. "ret" is in v0 at this point, but doing an
+    // explicit return uses an additional register.
+    // return ret;
```

Comment:
This game is so ducktaped together. I'm surprised it even works

Reply:
OOT is going to wrap that return in an `#ifdef AVOID_UB` for GCC. Maybe MM would want to do the same

Reply:
I think having it in is fine, but I think I'd prefer to wait until it's in the OoT repo incase there's any changes to the actual implementation of this in the review process. Once that's in though, I can apply it here